cds-snc/covid-alert-portal
# Search improvements: Partial word search and ignore accents

# Summary | Résumé

Improvements to the Location search rolled out in #500 

The Django full-text search function wrappers around the Postgres full-text search features are limited in that they do not support partial word search syntax. We also had an issue with french accents not being ignored for search purposes.

This PR refactors the search code to:
- Dynamically build up the search query with appropriate wildcard markup
- Use the underlying Postgres functions directly
- Ignore french accents for the purpose of search, so "Montreal" will return results including "Montréal"

A search query containing three words: 
- `first second third` 

would result in the following search query: 
- `first:* & second:* & third:*`

So, each string part is treated as a starts_with wildcard search, and are combined with AND operators.

[Based on idea described here](https://www.fusionbox.com/blog/detail/partial-word-search-with-postgres-full-text-search-in-django/632/)

## To test
Before checking out this branch, try some partial word searches and note that there are no results. Also if you have a location with the city "Montréal", try searching for "Montreal" (no accent) and note there are no results.

Then checkout this branch and try both of those searches again.

## Important note
To ignore accents, had to enable an Unaccent Postgres extension through a migration. I am uncertain if this migration will actually run on Staging/Production. Something to watch for when merging to Staging.
